[BLOG} Semantic Search (#31)
microsoft/Windows-DevRel: Blogs/GallerySampleSpotlight/SemanticSearch/blog.md

# Semantic Search

This blog post is the second in a series spotlighting the local AI samples contained in the [AI Dev Gallery.]() The Gallery is a preview project that aims to showcase local AI scenarios on Windows and to give developers the guidance they need to enable those scenarios themselves. The Gallery is open-source and contains a wide selection of different models and samples, including text, image, audio, and video use cases. The first blog post in the series, on Pose Estimation, can be found [here.]()

This second post will cover how to implement a semantic search with a locally running text embedding model. The Gallery currently includes the `all-MiniLM-L6-v2` and `all-MiniLM-L12-v2` models for text embedding tasks, and these models are run in the app via ONNX Runtime.

## Semantic Search At A High Level

> At its core, a semantic search is a search that matches meaning rather than directly matching text or text similarity. For example, if a user was looking for search results on the query term "sports," the source text only contained the word "athletics," a traditional search would find no matches, while a semantic search would surface the relevant text results. Semantic matching enables a robust version of searching that more closely parallels a human view of information, taking into account meaning, context, and nuance.

Reviewer: recommend to change "meaning" to be italic

recommend to simplify "For example, if a user was looking for search results on the query term "sports," the source text only contained the word "athletics," " to "For example, if a user did a search with the query term "sports," and the source text only contained the word "athletics,"

### How do text embedding models work?

> Text embedding models are the magic behind semantic search. These models convert source text to a numerical representation of the meaning of that original text. Instead of a string, you will have a vector of numerical values that is supposed to capture as much of the original meaning of the text as possible. 

Reviewer: recommend to include an image or two to help readers who are visual to understand

Author: ```suggestion
Text embedding models are the magic behind semantic search. These models convert source text to a numerical representation of the meaning of that original text. Instead of a string, you will have a vector of numerical values that captures as much of the original meaning of the text as possible. 
```

> For different pieces of text, these vectors will have the same dimensionality (or length) which makes comparing the distance between the two vectors relatively easy. The larger the distance between two vector representations, the farther apart in meaning the source texts supposedly are. And the inverse is true as well: smaller distance between embedding vectors means closer meaning.

Reviewer 2: There is a bit of inconsistency in this section in my opinion. 
Here you say something is _supposed_ to work a certain way:
> you will have a vector of numerical values that is supposed to capture as much of the original meaning of the text as possible.

Here you say again, _supposedly_ it works this way:
> the farther apart in meaning the source texts supposedly are

But then here its stated definitively:
> And the inverse is true as well: smaller distance between embedding vectors means closer meaning.

Reviewer: +1

Author: ```suggestion
For different pieces of text, these vectors will have the same dimensionality (or length) which makes comparing the distance between the two vectors relatively easy. The larger the distance between two vector representations, the farther apart in embedded meaning the two texts are. And the inverse is true as well: smaller distance between embedding vectors means closer meaning.
```

> Text embedding models aren't perfect, and have limitations. For example, they can usually only handle text of up to a specified length. For longer pieces of text, the input must be chunked into smaller pieces that are manageable by the model.

Reviewer 2: When I think of "not perfect" I think of getting inconsistent results but the next two things you say are just limitations.

Author: ```suggestion
Text embedding models do have some key limitations. For example, they can usually only handle text of up to a specified length. For longer pieces of text, the input must be chunked into smaller pieces that are manageable by the model.
```

> They also are often limited to capturing meaning for a single token in a sequence (like a sentence or paragraph), rather than the entire sequence. Because of this, sometime some extra processing is required to get embeddings that represent larger groupings of words.

Reviewer: could you provide more context on what is considered  "larger groupings of words" It would be nice to understand the scale, maybe give an example of what the volume of text that Text embedding models are ideal for and what is considered large that would require additional processing

Author: ```suggestion
They are also often limited to capturing meaning for a single token in a sequence, rather than the entire sequence. Because of this, sometime some extra processing is required to get embeddings that represent larger groupings of words like sentences or paragraphs.
```

### Enabling Semantic Search with Text Embedding

> If we can capture semantic meaning with a text embedding model as described above, our step-by-step for semantic search becomes relatively straightforward. Let's assume we already have a `sourceText` string and a `searchQuery` string:

Author: ```suggestion
Given that we have a model that can produce embeddings for input text, the algorithm for semantic search is fairly straightforward. Let's assume we already have a `sourceText` string and a `searchQuery` string:
```

1. Assuming `sourceText` is on the longer side, it must be broken it up into small enough pieces that the embedding model can handle.
1. Once the input has been chunked, both `sourceText` and `searchQuery` must first be tokenized before they can be passed to the embeddings model. In this case, the BERT tokenizer is used.
1. The tokenized input is passed to an ONNX inference session that runs the embeddings model with the desired inputs.
1. The text embeddings are output by the inference session.
1. These are token-by-token embeddings, so the embeddings for each sequence must be averaged to get an embedding for the entire string.
1. Now that `sourceText` and `searchQuery` have been embedded, a search can be performed to find the chunks of text from `sourceText` that most closely match the meaning of the `searchQuery`. This is done by calculating the distance between two embeddings.

> 1. The closest matches are output as semantic matches for the `searchQuery`.

Reviewer 2: How do you get back to text from the embeddings?

These steps will be broken down in more detail in the code walkthrough below. The full code can also be viewed alongside the sample [within the AI Dev Gallery](), or in the [GitHub repository]().

> ## Code Walkthrough

Reviewer: The code walkthrough has a lot and it would have been helpful for me to follow along if you used the steps you outlined on lines 23 -28 as the framework for the code walkthrough

* **AttentionMask:** The attention mask is an array of values of either 1 or 0 that tells the embedding model which tokens to pay attention to. Tokens that are assigned 1 are processed while tokens that are assigned 0 are ignored. This is necessary because the tokenizer encodes all strings in a batch to be of equal length, and will pad shorter strings to match the lengths. The attention mask tells the embedding model to ignore the padding tokens.
* **TokenTypes:** The token types array assigns "types" to each token in the input. For certain use cases, this can allow separator tokens to be denoted as different than normal input tokens. For now, this can mostly be ignored as it doesn't come up much when just dealing with plain strings.

Now that the input is flattened, it needs to converted to `OrtValues` so that it matches ONNX's expected input format:

```c#
    // Get the length of each of the input sequences by dividing the total length by the count
    int sequenceLength = input.InputIds.Length / count;

>     // Create input tensors over the input data using count and sequence length to define the shape

Reviewer: not sure if `tensors` is a common known word, if you think so great (I had to look it up and https://www.w3schools.com/ai/ai_tensors.asp is a great resource)

```C#
// Call to the vectorized search via the embeddings collection
VectorSearchResults<StringData> vectorSearchResults = await embeddingsCollection.VectorizedSearchAsync(
    searchVectors[0].Vector, // Pass in the search vector
    new VectorSearchOptions<StringData> // Pass in search options
    {
        Top = 5, // Number of results to return
        VectorProperty = (str) => str.Vector // The mapping to the vector property to compare on
    },
    ct).ConfigureAwait(false);
```

This search will return the 5 closest (or most similar in meaning) embedding vectors to the search vector stored in `searchVectors[0].Vector`. The closest vectors are found using a regular distance formula, but with 384 dimensions for each embedding value isntead of 2 or 3 as in real space.

The output is returned as a `VectorSearchResults<StringData>` and the final output strings can be accessed like so:

```C#
var resultMessage = string.Join("\n\n", vectorSearchResults.Results.ToBlockingEnumerable().Select(r => r.Record.Text));
```

Woohoo! That's all it takes to implement an effecive semantic search with a local text embedding model. 

> This walkthrough left out a lot of the extra code in the sample for updating the UI / error handling, but you can see the full sample code and try it out yourself in the [AI Dev Gallery]().

Reviewer: recommend to rephrase: 
This walkthrough was focused on generating text embeddings and using them to execute a semantic search, but there are other elements to factor in like error handling and updating the UI. You can see the full sample code....."

Author: ```suggestion
This walkthrough was focused on generating text embeddings and using them to execute a semantic search, but there are other elements to factor in like error handling and updating the UI. You can see the full sample code and try it out yourself in the [AI Dev Gallery]().
```

## Next Steps

> If you enjoyed this post and want to see more of what's possible with local AI on Windows, check out the [AI Dev Gallery]().

Reviewer: "If you enjoyed post like this one and want...
